src/lib.rs

```diff
@@ -43,34 +43,91 @@ impl Sequence {
-    pub fn seek(&mut self, sample: u64) {
+    /// Seeks to the given sample offset. Starts at dimension 0 of the given sample.
+    pub fn seek_to(&mut self, sample: u64) {
         self.sample = sample;
         self.dimension = 0;
     }
+    
+    /// Seeks a relative offset of samples. Starts at dimension 0 of the sample which is seeked to.
+    /// Performs a wrapping addition.
+    pub fn seek(&mut self, offset: u64) {
+        self.sample.wrapping_add(offset);
+        self.dimension = 0;
+    }
+    
+    /// Seeks to the given dimension within the current sample.
+    pub fn seek_to_dimension(&mut self, dimension: usize) {
+        assert!(dimension < self.dimensionality);
+        self.dimension = dimension;
+    }
+    
+    /// Seeks a relative offset of dimensions within the current sample.
+    pub fn seek_dimension(&mut self, offset: usize) {
+        let dimension = self.dimension + offset;
+        assert!(dimension < self.dimensionality);
+        self.dimension = dimension;
+    }
 
     /// Generates the next sample dimension as a [0, 1) fixed-point value.
     pub fn next_u64(&mut self) -> u64 {
-        (1u64 << 63).wrapping_add(self.next_raw())
+        raw_to_u64(self.next_raw())
     }
 
     /// Generates the next sample dimension as a [0, 1) fixed-point value.
     pub fn next_u32(&mut self) -> u32 {
-        (1u32 << 31).wrapping_add(Self::shift_round(self.next_raw(), 32) as u32)
+        raw_to_u32(self.next_raw())
     }
 
     /// Generates the next sample dimension as a [0, 1) floating-point value.
     pub fn next_f64(&mut self) -> f64 {
-        (0.5 + f64::from_bits(0x3FF0_0000_0000_0000 | Self::shift_round(self.next_raw(), 12)))
-            .fract()
+        raw_to_f64(self.next_raw())
     }
 
     /// Generates the next sample dimension as a [0, 1) floating-point value.
     pub fn next_f32(&mut self) -> f32 {
-        (0.5 + f32::from_bits(0x3F80_0000 | Self::shift_round(self.next_raw(), 41) as u32)).fract()
+        raw_to_f32(self.next_raw())
     }
-
-    fn next_raw(&mut self) -> u64 {
+    
+    /// Fills the slice with the next sample dimensions in the sequence. Will wrap to dimensions of the next sample
+    /// after the current sample has been exhausted.
+    pub fn fill_with_samples_f32(&mut self, samples: &mut [f32]) {
+        for v in samples.iter_mut() {
+            *v = self.next_f32();
+        }
+    }
+    
+    /// Fills the slice with the next sample dimensions in the sequence. Will wrap to dimensions of the next sample
+    /// after the current sample has been exhausted.
+    pub fn fill_with_samples_f64(&mut self, samples: &mut [f64]) {
+        for v in samples.iter_mut() {
+            *v = self.next_f64();
+        }
+    }
+    
+    /// Fills the slice with the next sample dimensions in the sequence. Will wrap to dimensions of the next sample
+    /// after the current sample has been exhausted.
+    pub fn fill_with_samples_u32(&mut self, samples: &mut [u32]) {
```

Comment:
I like this API but do you think it makes sense to fill "beyond" a full sample, wrapping around to the next one? I guess there isn't really anything more meaningful to do anyway...

The fill methods could probably be optimized a bit to avoid bumping the dimension index repeatedly, but meh that's an implementation detail.

Reply:
I think filling beyond a full sample is fine personally